plugins/inputs/docker_log/docker_logs.go

```diff
@@ -0,0 +1,440 @@
+		container.Close()
+	}
+	return nil
+}
+
+func (d *DockerLogs) containerListUpdate(acc telegraf.Accumulator) error {
+	ctx, cancel := context.WithTimeout(context.Background(), d.Timeout.Duration)
+	defer cancel()
+	if d.client == nil {
+		log.Printf("%s : Dock client is null", ERR_PREFIX)
+		return nil
+	}
+	containers, err := d.client.ContainerList(ctx, d.opts)
+	if err != nil {
+		log.Printf("%s : %s ", ERR_PREFIX, err.Error())
+		return err
+	}
+	for _, container := range containers {
+		if d.containerInContainerList(container.ID) {
+			continue
+		}
+		d.wg.Add(1)
+		/*Start a new goroutine for every new container that has logs to collect*/
+		go func(c types.Container) {
+			defer d.wg.Done()
+			err := d.getContainerLogs(c, acc)
+			if err != nil {
+				log.Printf("%s : %s ", ERR_PREFIX, err.Error())
+			}
+			d.removeFromContainerList(c.ID)
+			return
+		}(container)
+	}
+	return nil
+}
+
+func (d *DockerLogs) getContainerLogs(
+	container types.Container,
+	acc telegraf.Accumulator,
+) error {
+	c, err := d.client.ContainerInspect(context.Background(), container.ID)
+	if err != nil {
+		log.Printf("%s : %s", ERR_PREFIX, err.Error())
+		return err
+	}
+	logOptions := types.ContainerLogsOptions{
+		ShowStdout: true,
+		ShowStderr: true,
+		Timestamps: false,
+		Details:    true,
+		Follow:     true,
+		Tail:       "0",
+	}
+	logReader, err := d.client.ContainerLogs(context.Background(), container.ID, logOptions)
+	d.addToContainerList(container.ID, logReader)
+	if err != nil {
+		log.Printf("%s : %s", ERR_PREFIX, err.Error())
+		return err
+	}
+	/* Parse container name */
+	cname := "unknown"
+	if len(container.Names) > 0 {
+		/*Pick first container name*/
+		cname = strings.TrimPrefix(container.Names[0], "/")
+	}
+
+	tags := map[string]string{
+		"containerId":   container.ID,
+		"containerName": cname,
+	}
+	// Add labels to tags
+	for k, label := range container.Labels {
+		if d.labelFilter.Match(k) {
+			tags[k] = label
+		}
+	}
+	if c.Config.Tty {
+		_, err = pushTtyLogs(acc, tags, logReader)
+	} else {
+		_, err = pushLogs(acc, tags, logReader)
+	}
+	if err != nil {
+		log.Printf("%s : %s", ERR_PREFIX, err.Error())
+		return err
+	}
+	return nil
+}
+func pushTtyLogs(acc telegraf.Accumulator, tags map[string]string, src io.Reader) (written int64, err error) {
+	tags["logType"] = "unknown" //in tty mode we wont be able to differentiate b/w stdout and stderr hence unknown
+	fields := map[string]interface{}{}
+	data := make([]byte, LOG_BYTES_MAX)
+	for {
+		num, err := src.Read(data)
+		if err != nil {
+			if err == io.EOF {
+				written += int64(num)
+				fields["log"] = data[1:num]
+				acc.AddFields("docker_log", fields, tags)
+				return written, nil
+			}
+			return written, err
+		}
+		written += int64(num)
+		if len(data) > 0 {
+			fields["log"] = data[1:num]
+			acc.AddFields("docker_log", fields, tags)
+		}
+	}
+}
+
+/* Inspired from https://github.com/moby/moby/blob/master/pkg/stdcopy/stdcopy.go */
```

Comment:
It seems like an additional overhead and complexity . It would be additional goroutines per docker. 

By 2 goroutines you mean one to run stdcopy (since it would be blocking) and another to read from dsterr,dstout and throw it to telegraf acc ? 

I'm not entirely sure how to approach this!

Reply:
Don't worry about it for now.